src/test/function_test/test_throttle.cpp

```diff
@@ -0,0 +1,790 @@
+{
+    get,
+    multi_get,
+    set,
+    multi_set
+};
+
+struct throttle_test_plan
+{
+    std::string test_plan_case;
+    operation_type ot;
+    int single_value_sz;
+    int multi_count;
+    int limit_qps;
+    bool random_value_size = false;
+    bool is_hotkey = false;
+};
+
+#define ToString(x) #x
+
+#define TIMELY_RECORD(time_interval, is_reject, size)                                              \
+    do {                                                                                           \
+        records[ToString(time_interval##_query_times)]++;                                          \
+        records[ToString(time_interval##_query_size)] += size;                                     \
+        if (is_reject) {                                                                           \
+            records[ToString(time_interval##_reject_times)]++;                                     \
+            records[ToString(time_interval##_reject_size)] += size;                                \
+        } else {                                                                                   \
+            records[ToString(time_interval##_successful_times)]++;                                 \
+            records[ToString(time_interval##_successful_size)] += size;                            \
+        }                                                                                          \
+    } while (0)
+
+struct throttle_test_recorder
+{
+    uint64_t start_time_ms;
+    uint64_t duration_ms;
+    std::map<std::string, uint64_t> records;
+    std::string test_name;
+    std::vector<std::string> parameter_seq = {"total_qps",
+                                              "total_size_per_sec",
+                                              "first_10_ms_successful_times",
+                                              "first_100_ms_successful_times",
+                                              "first_1000_ms_successful_times",
+                                              "first_5000_ms_successful_times",
+                                              "first_10_ms_successful_size",
+                                              "first_100_ms_successful_size",
+                                              "first_1000_ms_successful_size",
+                                              "first_5000_ms_successful_size"};
+
+    void start_test(const std::string &test_case, uint64_t time_duration_s)
+    {
+        test_name = test_case;
+        start_time_ms = dsn_now_ms();
+        duration_ms = time_duration_s * 1000;
+        records.emplace(std::make_pair("duration_ms", duration_ms));
+    }
+
+    bool is_time_up()
+    {
+        if (dsn_now_ms() - start_time_ms > duration_ms) {
+            return true;
+        }
+        return false;
+    }
+
+    void record(uint64_t size, bool is_reject)
+    {
+        if (is_time_up()) {
+            return;
+        }
+        if (dsn_now_ms() - start_time_ms <= 10) {
```

Comment:
```
switch(now_ms - start_time_ms) {
case 0...10:
case 11...100:
```

Reply:
I think it's just code style differences, no need to change

Reply:
I don't agree with you. Writing too many `if...else` is bad smell

Reply:
switch in C++ can't support `>` or `<`